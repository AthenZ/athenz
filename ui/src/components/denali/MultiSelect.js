/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';
import { css, cx } from '@emotion/css';

const makeCssMultiSelect = () => css`
    .denali-multiselect__control {
        min-height: 36px !important;
        width: auto;
        padding: 0px 10px !important;
    }

    & > .denali-multiselect__control {
        align-items: center;
        background: rgba(1, 110, 255, 0.08) !important;
        border: none !important;
        border-radius: 4px;
        color: #303030;
        display: flex;
        outline: none;
        position: relative;
    }

    .denali-multiselect__control--menu-is-open {
        box-shadow: 0px 2px 0px #016eff !important;
    }

    & > .denali-multiselect__control--is-focused {
        align-items: center;
        background: #016eff14 !important;
        border: none !important;
        border-radius: 4px;
        color: #303030;
        display: flex;
        min-height: 36px;
        outline: none;
        padding: 2px 25px 2px 10px;
        position: relative;
    }

    .denali-multiselect__value-container {
        padding: 0% !important;
    }

    & > .denali-multiselect__control--menu-is-open {
        align-items: center;
        background: rgba(1, 110, 255, 0.08) !important;
        border: none !important;
        border-radius: 4px;
        color: #303030;
        display: flex;
        min-height: 36px;
        outline: none;
        padding: 2px 25px 2px 10px;
        position: relative;
    }

    .denali-multiselect__multi-value__label {
        font-size: 1.4rem !important;
        padding: 0 !important;
        padding-left: 7px !important;
        margin: 0;
    }

    /* tag remove icon */
    .denali-multiselect__multi-value__remove {
        content: '';
        height: 20px;
        width: 20px;
        background-image: url('/static/close.svg');
        background-repeat: no-repeat;
        background-size: 14px;
        margin-top: 6px;
        margin-left: 5px;
    }

    .denali-multiselect__multi-value__remove > svg {
        display: none;
    }

    .denali-multiselect__multi-value__remove:hover {
        background-color: transparent !important;
        cursor: pointer;
    }

    /* drop down arrowhead */
    .denali-multiselect__dropdown-indicator {
        content: '';
        right: 10px;
    }

    /* clear all indicator try after or before and try hiding svg*/
    .denali-multiselect__clear-indicator > svg {
        display: none;
    }

    .denali-multiselect__clear-indicator:before {
        content: 'Clear all';
    }

    .denali-multiselect__clear-indicator {
        display: block;
        color: #016eff !important;
        text-decoration: none;
        font-size: inherit;
        cursor: pointer;
    }

    .denali-multiselect__clear-indicator:hover {
        color: #004cb3 !important;
    }

    .denali-multiselect__menu-list {
        padding: 0 !important;
        font-size: 1.4rem;
        font-family: Helvetica, 'Arial', sans-serif;
    }

    .denali-multiselect__control {
        align-items: center;
        background: rgba(1, 110, 255, 0.08) !important;
        border: none !important;
        border-radius: 4px;
        box-shadow: none;
        color: #303030;
        display: flex;
        min-height: 36px;
        outline: none;
        padding: 2px 25px 2px 10px;
        position: relative;
    }

    .denali-multiselect__indicator-separator {
        display: none !important;
    }

    .denali-multiselect__multi-value {
        display: inline-flex;
        align-items: center;
        justify-content: center;
        padding: 0px 10px;
        height: 28px;
        margin: 3px;
        background-color: #cce2ff !important;
        color: #303030;
        border-radius: 9999px !important;
        white-space: nowrap;
        transition: 300ms;
        font-size: 1.2rem;
        font-family: Helvetica, 'Arial', sans-serif;
        height: 22px;
        padding: 0px 8px;
    }

    .denali-multiselect__multi-value__remove {
        padding: 0 !important;
    }

    .denali-multiselect__multi-value__remove :hover {
        background-color: #cce2ff !important;
        color: #3d4042 !important;
        cursor: pointer;
    }
`;

class MultiSelect extends React.PureComponent {
    render() {
        const {
            name,
            closeMenuOnSelect,
            isClearable,
            isSearchable,
            options,
            selectedValues,
            onChange,
            disabled,
            placeholder,
        } = this.props;

        let cssMultiSelect = makeCssMultiSelect();
        const classes = cx(cssMultiSelect, 'denali-multiselect');

        const authorityFilters = selectedValues.split(',').map((filter) => {
            return {
                label: filter,
                value: filter,
            };
        });

        return (
            <div data-testid='denali-multiselect'>
                <Select
                    defaultValue={selectedValues && authorityFilters}
                    className={classes}
                    isMulti
                    name={name}
                    options={options.map((option) => {
                        return {
                            label: option.value,
                            value: option.name,
                        };
                    })}
                    classNamePrefix='denali-multiselect'
                    closeMenuOnSelect={closeMenuOnSelect}
                    isClearable={isClearable}
                    isSearchable={isSearchable}
                    onChange={onChange}
                    isDisabled={disabled}
                    placeholder={placeholder}
                />
            </div>
        );
    }
}

MultiSelect.propTypes = {
    name: PropTypes.string,
    options: PropTypes.arrayOf(
        PropTypes.shape({
            label: PropTypes.string,
            value: PropTypes.string,
        })
    ).isRequired,
    closeMenuOnSelect: PropTypes.bool,
    isClearable: PropTypes.bool,
    isSearchable: PropTypes.bool,
};

export default MultiSelect;
