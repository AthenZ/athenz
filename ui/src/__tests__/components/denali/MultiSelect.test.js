/*
 * Copyright The Athenz Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import MultiSelect from '../../../components/denali/MultiSelect';

describe('MultiSelect', () => {
    const CLASSNAME_PREFIX = '.denali-multiselect';
    const AUTH_FILTER_ONSHORE = 'OnShore-US';
    const AUTH_FILTER_DATA_GOVERNANCE = 'DataGovernance';

    const onChange = jest.fn();

    afterEach(() => jest.clearAllMocks());

    const options = [
        { label: 'OnShore-US', value: 'OnShore-US' },
        { label: 'DataGovernance', value: 'DataGovernance' },
    ];

    it('renders initial (closed) state', () => {
        const { baseElement, getByTestId } = render(
            <MultiSelect
                options={options}
                onChange={onChange}
                selectedValues={''}
            />
        );

        const multiselect = getByTestId('denali-multiselect');

        expect(
            baseElement.querySelector('.denali-multiselect__menu')
        ).toBeNull();

        expect(
            baseElement.querySelectorAll('.denali-multiselect__option')
        ).toHaveLength(0);

        expect(multiselect).toMatchSnapshot();
    });

    it('renders opened multiselect dropdown', async () => {
        const { baseElement, getByTestId } = render(
            <MultiSelect
                options={options}
                onChange={onChange}
                selectedValues={''}
            />
        );

        const multiselect = baseElement.querySelector(
            `${CLASSNAME_PREFIX}__control`
        );
        fireEvent.mouseDown(multiselect);

        await waitFor(() =>
            baseElement.querySelector(`${CLASSNAME_PREFIX}__menu`)
        );

        const dropdownItems = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__option`
        );

        expect(dropdownItems).toHaveLength(options.length);
    });

    it('renders with default selected item', () => {
        const { baseElement, getByTestId } = render(
            <MultiSelect
                options={options}
                onChange={onChange}
                selectedValues={AUTH_FILTER_ONSHORE}
            />
        );

        const selectedOptions = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__multi-value__label`
        );

        expect(selectedOptions.length).toBe(1);
        expect(selectedOptions[0].textContent).toEqual(AUTH_FILTER_ONSHORE);
    });

    it('can remove a selected item', async () => {
        const { baseElement } = render(
            <MultiSelect
                options={options}
                onChange={onChange}
                selectedValues={`${AUTH_FILTER_ONSHORE},${AUTH_FILTER_DATA_GOVERNANCE}`}
            />
        );

        let selectedOptions = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__multi-value__label`
        );
        const removeButtons = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__multi-value__remove`
        );

        expect(selectedOptions).toHaveLength(options.length);
        expect(removeButtons).toHaveLength(options.length);
        expect(selectedOptions[0].textContent).toEqual(AUTH_FILTER_ONSHORE);
        expect(selectedOptions[1].textContent).toEqual(
            AUTH_FILTER_DATA_GOVERNANCE
        );

        fireEvent.click(removeButtons[1]);

        await waitFor(() => {
            expect(selectedOptions[1]).not.toBeInTheDocument();
        });

        selectedOptions = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__multi-value__label`
        );

        expect(selectedOptions).toHaveLength(1);
        expect(selectedOptions[0].textContent).toEqual(AUTH_FILTER_ONSHORE);
    });

    it('can open the multiselect menu and select an item', async () => {
        const { baseElement, getByTestId } = render(
            <MultiSelect
                options={options}
                onChange={onChange}
                selectedValues={''}
            />
        );

        const multiselect = baseElement.querySelector(
            `${CLASSNAME_PREFIX}__control`
        );
        fireEvent.mouseDown(multiselect);

        await waitFor(() =>
            baseElement.querySelector(`${CLASSNAME_PREFIX}__menu`)
        );

        const dropdownItem = baseElement.querySelector(
            `${CLASSNAME_PREFIX}__option`
        );
        fireEvent.click(dropdownItem);

        const selectedOptions = baseElement.querySelectorAll(
            `${CLASSNAME_PREFIX}__multi-value__label`
        );

        expect(selectedOptions).toHaveLength(1);
    });
});
