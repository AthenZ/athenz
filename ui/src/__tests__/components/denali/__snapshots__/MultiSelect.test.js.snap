// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect renders initial (closed) state 1`] = `
.emotion-0 .denali-multiselect__control {
  min-height: 36px!important;
  width: auto;
  padding: 0px 10px!important;
}

.emotion-0>.denali-multiselect__control {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(1, 110, 255, 0.08)!important;
  border: none!important;
  border-radius: 4px;
  color: #303030;
  display: flex;
  outline: none;
  position: relative;
}

.emotion-0 .denali-multiselect__control--menu-is-open {
  box-shadow: 0px 2px 0px #016eff!important;
}

.emotion-0>.denali-multiselect__control--is-focused {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #016eff14!important;
  border: none!important;
  border-radius: 4px;
  color: #303030;
  display: flex;
  min-height: 36px;
  outline: none;
  padding: 2px 25px 2px 10px;
  position: relative;
}

.emotion-0 .denali-multiselect__value-container {
  padding: 0%!important;
}

.emotion-0>.denali-multiselect__control--menu-is-open {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(1, 110, 255, 0.08)!important;
  border: none!important;
  border-radius: 4px;
  color: #303030;
  display: flex;
  min-height: 36px;
  outline: none;
  padding: 2px 25px 2px 10px;
  position: relative;
}

.emotion-0 .denali-multiselect__multi-value__label {
  font-size: 1.4rem!important;
  padding: 0!important;
  padding-left: 7px!important;
  margin: 0;
}

.emotion-0 .denali-multiselect__multi-value__remove {
  content: '';
  height: 20px;
  width: 20px;
  background-image: url('/static/close.svg');
  background-repeat: no-repeat;
  -webkit-background-size: 14px;
  background-size: 14px;
  margin-top: 6px;
  margin-left: 5px;
}

.emotion-0 .denali-multiselect__multi-value__remove>svg {
  display: none;
}

.emotion-0 .denali-multiselect__multi-value__remove:hover {
  background-color: transparent!important;
  cursor: pointer;
}

.emotion-0 .denali-multiselect__dropdown-indicator {
  content: '';
  right: 10px;
}

.emotion-0 .denali-multiselect__clear-indicator>svg {
  display: none;
}

.emotion-0 .denali-multiselect__clear-indicator:before {
  content: 'Clear all';
}

.emotion-0 .denali-multiselect__clear-indicator {
  display: block;
  color: #016eff!important;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-size: inherit;
  cursor: pointer;
}

.emotion-0 .denali-multiselect__clear-indicator:hover {
  color: #004cb3!important;
}

.emotion-0 .denali-multiselect__menu-list {
  padding: 0!important;
  font-size: 1.4rem;
  font-family: Helvetica,'Arial',sans-serif;
}

.emotion-0 .denali-multiselect__control {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: rgba(1, 110, 255, 0.08)!important;
  border: none!important;
  border-radius: 4px;
  box-shadow: none;
  color: #303030;
  display: flex;
  min-height: 36px;
  outline: none;
  padding: 2px 25px 2px 10px;
  position: relative;
}

.emotion-0 .denali-multiselect__indicator-separator {
  display: none!important;
}

.emotion-0 .denali-multiselect__multi-value {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  padding: 0px 8px;
  height: 22px;
  margin: 3px;
  background-color: #cce2ff!important;
  color: #303030;
  border-radius: 9999px!important;
  white-space: nowrap;
  -webkit-transition: 300ms;
  transition: 300ms;
  font-size: 1.2rem;
  font-family: Helvetica,'Arial',sans-serif;
}

.emotion-0 .denali-multiselect__multi-value__remove {
  padding: 0!important;
}

.emotion-0 .denali-multiselect__multi-value__remove :hover {
  background-color: #cce2ff!important;
  color: #3d4042!important;
  cursor: pointer;
}

.emotion-1 {
  position: relative;
  box-sizing: border-box;
}

.emotion-2 {
  z-index: 9999;
  border: 0;
  clip: rect(1px, 1px, 1px, 1px);
  height: 1px;
  width: 1px;
  position: absolute;
  overflow: hidden;
  padding: 0;
  white-space: nowrap;
}

.emotion-4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: default;
  display: flex;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-height: 38px;
  outline: 0!important;
  position: relative;
  -webkit-transition: all 100ms;
  transition: all 100ms;
  background-color: hsl(0, 0%, 100%);
  border-color: hsl(0, 0%, 80%);
  border-radius: 4px;
  border-style: solid;
  border-width: 1px;
  box-sizing: border-box;
}

.emotion-4:hover {
  border-color: hsl(0, 0%, 70%);
}

.emotion-5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: grid;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-box-flex-wrap: wrap;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-overflow-scrolling: touch;
  position: relative;
  overflow: hidden;
  padding: 2px 8px;
  box-sizing: border-box;
}

.emotion-6 {
  grid-area: 1/1/2/3;
  color: hsl(0, 0%, 50%);
  margin-left: 2px;
  margin-right: 2px;
  box-sizing: border-box;
}

.emotion-7 {
  visibility: visible;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  display: inline-grid;
  grid-area: 1/1/2/3;
  grid-template-columns: 0 min-content;
  margin: 2px;
  padding-bottom: 2px;
  padding-top: 2px;
  color: hsl(0, 0%, 20%);
  box-sizing: border-box;
}

.emotion-7:after {
  content: attr(data-value) " ";
  visibility: hidden;
  white-space: pre;
  grid-area: 1/2;
  font: inherit;
  min-width: 2px;
  border: 0;
  margin: 0;
  outline: 0;
  padding: 0;
}

.emotion-8 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  box-sizing: border-box;
}

.emotion-9 {
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  width: 1px;
  background-color: hsl(0, 0%, 80%);
  margin-bottom: 8px;
  margin-top: 8px;
  box-sizing: border-box;
}

.emotion-10 {
  display: flex;
  -webkit-transition: color 150ms;
  transition: color 150ms;
  color: hsl(0, 0%, 80%);
  padding: 8px;
  box-sizing: border-box;
}

.emotion-10:hover {
  color: hsl(0, 0%, 60%);
}

.emotion-11 {
  display: inline-block;
  fill: currentColor;
  line-height: 1;
  stroke: currentColor;
  stroke-width: 0;
}

<div
  data-testid="denali-multiselect"
>
  <div
    class="emotion-0 denali-multiselect emotion-1"
  >
    <span
      class="emotion-2"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="emotion-2"
      role="log"
    />
    <div
      class="denali-multiselect__control emotion-4"
    >
      <div
        class="denali-multiselect__value-container denali-multiselect__value-container--is-multi emotion-5"
      >
        <div
          class="denali-multiselect__placeholder emotion-6"
          id="react-select-2-placeholder"
        >
          Select...
        </div>
        <div
          class="denali-multiselect__input-container emotion-7"
          data-value=""
        >
          <input
            aria-activedescendant=""
            aria-autocomplete="list"
            aria-describedby="react-select-2-placeholder"
            aria-expanded="false"
            aria-haspopup="true"
            autocapitalize="none"
            autocomplete="off"
            autocorrect="off"
            class="denali-multiselect__input"
            id="react-select-2-input"
            role="combobox"
            spellcheck="false"
            style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
            tabindex="0"
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        class="denali-multiselect__indicators emotion-8"
      >
        <span
          class="denali-multiselect__indicator-separator emotion-9"
        />
        <div
          aria-hidden="true"
          class="denali-multiselect__indicator denali-multiselect__dropdown-indicator emotion-10"
        >
          <svg
            aria-hidden="true"
            class="emotion-11"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;
